### PR TITLE
Introduce functionality for chunking and breaking IID experiments

### PR DESCRIPTION
**Pull request by mcw92**
This PR introduces functionality for the chunking and breaking IID experiments. In particular, the evaluation has been extended to calculate and save local and global confusion matrices in order to enable calculation of arbitrary metrics for the breaking IID experiments. 

The following changes have been made:
- Make the synthetic data generation consistent throughout the code. This means that in the serial case, the dataset generated with `generate_and_distribute_synthetic_dataset` _without local or global imbalances_ equals the _completely balanced_ dataset generated with `make_classification_dataset` when using the same random state. This ensures comparability of the strong scaling experiment series with and without chunking as the same datasets are created when passing the same random state.
- Fix passing additional keyword arguments in both `train_parallel_on_synthetic data` and `train_parallel_on_balanced_synthetic_data`. This was completely missing in the former case. In addition, the argument parser was lacking some of the keyword arguments used in `sklearn`'s `make_classification` and `train_test_split` used under the hood.
- Introduce job script generation scripts for both chunking and breaking IID experiments.
- Add calculation and saving of local and global confusion matrices, including tests.
- Add evaluation from checkpoints for breaking IID experiments.
- Refactor `train` module into `train_serial` and `train_parallel`.
- Remove MacOS from build matrix used for the tests as this caused problems with tests hanging randomly forever for different Python versions in different test runs (see #19).

The plotting scripts are still kind of messy with many code redundancies. I will fix this in the future. For now, I would like to prioritize the things required to run the experiments. **If the PR is too messy, please just tell me 🙈.**

### Notes to self:
- `sklearn`'s `RandomForestClassifier` internally uses weighted voting in its [`predict()` method](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html#sklearn.ensemble.RandomForestClassifier.predict), i.e., the predicted class of an input sample is a vote by the trees in the forest, weighted by their probability estimates. The predicted class thus is the one with highest mean probability estimate across the trees. As the `DistributedRandomForest` class in `specialcouscous` only implements plain voting, I also implemented plain voting for calculation of the local confusion matrices instead of using `predict()` in order to ensure consistency and comparability.
- Possible problem with the confusion matrix might occur when the local data does not contain all classes for extremely imbalanced datasets / data partitioning. However, I am not sure about this.
- As building a globally shared model turned out to be infeasible for most of our use cases / experiments, the functionality for calculating the confusion matrix and also evaluating breaking IID experiments from checkpoints mainly focuses on the case where the global model is not shared but distributed. That is why a shared test set is required in all our experiments.

### CODE REVIEW

File: scripts/analysis/strong_scaling.py [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+pd.set_option("display.max_rows", None)
+
+
+if __name__ == "__main__":
+    # Get the root directory where results are stored from command line.
+    root_dir = sys.argv[1]
+    data_set = root_dir.split(os.sep)[-1]
+    # Dictionary to store the values grouped by (dataset, number_of_tasks, data seed)
+    results = defaultdict(list)
+    data_set = root_dir.split(os.sep)[-1]
```
Comment (fluegelk) [inline scripts/analysis/strong_scaling.py:30]:
```suggestion
    data_set = root_dir.split(os.sep)[-1]
    # Dictionary to store the values grouped by (dataset, number_of_tasks, data seed)
    results = defaultdict(list)
```
Duplicate definition of `data_set`?

**Review by fluegelk — APPROVED** [review]
Looks mostly good to me. I left a few minor comments which should be easy to fix. 

I did not go through the plotting scripts in `scripts/analysis` in detail, a lot of it can probably be deduplicated whenever we address issue #12.

File: specialcouscous/train/train_parallel.py [hunk 2]
```diff
@@ -25,186 +24,323 @@
...
+    n_classes: int,
+    globally_balanced: bool,
+    locally_balanced: bool,
+    shared_test_set: bool,
+    random_state: int | np.random.RandomState = 0,
+    random_state_model: int | None = None,
+    mu_partition: float | str | None = None,
+    mu_data: float | str | None = None,
+    peak: int | None = None,
+    make_classification_kwargs: dict[str, Any] = {},
```
[the remark this answers is not in the document]
Comment (mcw92) [reply specialcouscous/train/train_parallel.py:39]: Understood, I will change all default arguments back to `None`.

File: specialcouscous/synthetic_classification_data.py [hunk 3]
```diff
@@ -593,7 +593,7 @@ def generate(
         n_classes: int,
         class_weights: npt.NDArray[np.float32] | None = None,
         random_state: int | np.random.RandomState | None = None,
-        **kwargs: Any,
+        make_classification_kwargs: dict[str, Any] = {},
```
[the remark this answers is not in the document]
Comment (mcw92) [reply specialcouscous/synthetic_classification_data.py]: Fixed.

File: specialcouscous/synthetic_classification_data.py [hunk 4]
```diff
@@ -646,7 +647,7 @@ def generate_with_skellam_class_imbalance(
         peak: int = 0,
         rescale_to_sum_one: bool = True,
         random_state: int | np.random.RandomState | None = None,
-        **kwargs: Any,
+        make_classification_kwargs: dict[str, Any] = {},
```
[the remark this answers is not in the document]
Comment (mcw92) [reply specialcouscous/synthetic_classification_data.py]: Fixed.

File: specialcouscous/synthetic_classification_data.py [hunk 5]
```diff
@@ -897,10 +899,10 @@ def generate_and_distribute_synthetic_dataset(
     mu_data: float | str | None = None,
     peak: int | None = None,
     rescale_to_sum_one: bool = True,
-    make_classification_kwargs: dict | None = None,
+    make_classification_kwargs: dict[str, Any] = {},
```
[the remark this answers is not in the document]
Comment (mcw92) [reply specialcouscous/synthetic_classification_data.py]: Fixed.

File: specialcouscous/synthetic_classification_data.py [hunk 6]
```diff
@@ -1138,12 +1154,11 @@ def data_generation_demo(
 def make_classification_dataset(
     n_samples: int,
     n_features: int,
-    frac_informative: float = 0.1,
-    frac_redundant: float = 0.1,
     n_classes: int = 10,
-    n_clusters_per_class: int = 1,
+    make_classification_kwargs: dict[str, Any] = {},
```
[the remark this answers is not in the document]
Comment (mcw92) [reply specialcouscous/synthetic_classification_data.py]: Fixed.

File: specialcouscous/train/train_parallel.py [hunk 7]
```diff
@@ -245,208 +381,59 @@ def save_results_parallel(
...
 
-def train_parallel_on_synthetic_data(
+def train_parallel_on_balanced_synthetic_data(
     n_samples: int,
     n_features: int,
     n_classes: int,
-    globally_balanced: bool,
-    locally_balanced: bool,
-    shared_test_set: bool,
+    make_classification_kwargs: dict[str, Any] = {},
```
[the remark this answers is not in the document]
Comment (mcw92) [reply specialcouscous/train/train_parallel.py]: Fixed.

File: specialcouscous/train/train_parallel.py [hunk 8]
```diff
@@ -896,49 +874,62 @@ def train_parallel_on_balanced_synthetic_data(
...
     n_classes: int,
-    n_clusters_per_class: int,
-    frac_informative: float,
-    frac_redundant: float,
+    globally_balanced: bool,
+    locally_balanced: bool,
+    mu_partition: float | str | None = None,
+    mu_data: float | str | None = None,
+    peak: int | None = None,
+    make_classification_kwargs: dict[str, Any] = {},
```
[the remark this answers is not in the document]
Comment (mcw92) [reply specialcouscous/train/train_parallel.py]: Fixed.

File: specialcouscous/train/train_serial.py [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+log = logging.getLogger(__name__)  # Get logger instance.
+
+
+def train_serial_on_synthetic_data(
+    n_samples: int,
+    n_features: int,
+    n_classes: int,
+    random_state: int | np.random.RandomState = 0,
+    random_state_model: int | np.random.RandomState | None = None,
+    make_classification_kwargs: dict[str, Any] = {},
```
[the remark this answers is not in the document]
Comment (mcw92) [reply specialcouscous/train/train_serial.py:31]: Fixed.

File: specialcouscous/train/train_parallel.py [hunk 10]
```diff
@@ -506,204 +481,214 @@ def train_parallel_on_synthetic_data(
...
     n_features: int,
     n_classes: int,
-    n_clusters_per_class: int,
-    frac_informative: float,
-    frac_redundant: float,
+    make_classification_kwargs: dict[str, Any] = {},
     random_state: int | np.random.RandomState = 0,
+    checkpoint_path: str | pathlib.Path = pathlib.Path("../"),
+    checkpoint_uid: str = "",
     random_state_model: int | None = None,
```
[the remark this answers is not in the document]
Comment (mcw92) [reply specialcouscous/train/train_parallel.py:677]: You are right -- if I remember correctly, I introduced the model random state in these functions to make sure that I can really reproduce everything correctly from the checkpoints, including the `DistributedForest` object with its random state. However, I apparently did not pass the model random state further down to the actual `DistributedForest` instance which makes it kind of useless in the current form. I fixed this now. I agree that this is not really necessary and the model random state could also be removed completely; however, I would prefer to keep it for now for reasons of consistency in the function signatures and the `DistributedForest` object.

File: specialcouscous/train/train_parallel.py [hunk 11]
```diff
@@ -896,49 +874,62 @@ def train_parallel_on_balanced_synthetic_data(
...
+    locally_balanced: bool,
+    mu_partition: float | str | None = None,
+    mu_data: float | str | None = None,
+    peak: int | None = None,
+    make_classification_kwargs: dict[str, Any] = {},
     random_state: int | np.random.RandomState = 0,
-    checkpoint_path: str | pathlib.Path = pathlib.Path("./"),
+    checkpoint_path: str | pathlib.Path = pathlib.Path("../"),
     checkpoint_uid: str = "",
     random_state_model: int | None = None,
```
[the remark this answers is not in the document]
Comment (mcw92) [reply specialcouscous/train/train_parallel.py:919]: See comment above.
